packages/app/src/cli/models/app/app.ts

```diff
@@ -20,13 +20,215 @@ export const LegacyAppSchema = zod
+      if (topLevelDestination && topics.length) {
+        for (const topic of topics) {
+          const key = `${topic}${delimiter}${topLevelDestination}`
+
+          if (subscriptionDestinationsSet.has(key)) {
+            ctx.addIssue({
+              code: zod.ZodIssueCode.custom,
+              message: "You can't have duplicate subscriptions with the exact same topic and destination",
+              fatal: true,
+              path: ['topics', topic],
+            })
+            return zod.NEVER
+          }
+
+          subscriptionDestinationsSet.add(key)
+        }
+      }
+
+      // validate individual subscriptions
+      if (subscriptions.length) {
+        for (const [i, subscription] of subscriptions.entries()) {
+          const subscriptionDestinations = [
+            subscription.subscription_endpoint_url,
+            subscription.pubsub_project && subscription.pubsub_topic,
+            subscription.arn,
+          ].filter(Boolean)
+          const path = ['subscriptions', i]
+
+          if ([subscription.pubsub_project, subscription.pubsub_topic].filter(Boolean).length === 1) {
+            ctx.addIssue({
+              code: zod.ZodIssueCode.custom,
+              message: `${pubSubValidationErrMsg} a pub sub destination`,
+              fatal: true,
+              path,
+            })
+            return zod.NEVER
+          }
+
+          if (subscriptionDestinations.length > 1) {
+            ctx.addIssue({
+              code: zod.ZodIssueCode.custom,
+              message: `${tooManyDestinationsErrMsg} per subscription`,
+              fatal: true,
+              path,
+            })
+            return zod.NEVER
+          }
+
+          // If no top-level destinations are provided, ensure each subscription has at least one destination
+          if (!topLevelDestinations.length && subscriptionDestinations.length === 0) {
+            ctx.addIssue({
+              code: zod.ZodIssueCode.custom,
+              message: 'You must declare either a top-level destination or a destination per subscription',
+              fatal: true,
+              path,
+            })
+            return zod.NEVER
+          }
+
+          if (!subscription_endpoint_url && !subscription.subscription_endpoint_url && subscription.path) {
+            ctx.addIssue({
+              code: zod.ZodIssueCode.custom,
+              message: 'You must declare a subscription_endpoint_url if you wish to use a relative path',
+              fatal: true,
+              path,
+            })
+            return zod.NEVER
+          }
+
+          let destination = [
+            subscription.subscription_endpoint_url,
+            subscription.pubsub_project,
+            subscription.pubsub_topic,
+            subscription.arn,
+          ]
+            .filter(Boolean)
+            .join(delimiter)
+
+          // if there is no destination override, use top level destination
```

Comment:
it sounds like we're using `destination` to refer to an HTTPS endpoint, Google PubSub project+topic, or ARN

Reply:
This was my choice of vocabulary for exactly that, not sure if its appropriate or not but will get Jen's thoughts once the logic is finalized!

Reply:
100% agree, I think @jenstanicak will have some input, this was just my first ditch effort at finding a good word to describe everything